### PR TITLE
Rpc transaction counting

### PR DESCRIPTION
**Pull request by michalsidzej**
Resolves L2B-75

### CODE REVIEW

File: packages/backend/src/Application.ts [hunk 1]
```diff
@@ -59,15 +62,21 @@ export class Application {
...
+    const txCountRepository = new TxCountRepository(database, logger)
 
     const http = new HttpClient()
 
-    const provider = new providers.AlchemyProvider(
+    const ethereumProvider = new providers.AlchemyProvider(
       'mainnet',
       config.alchemyApiKey,
     )
+    const optimismProvider = new providers.AlchemyProvider(
```
Comment (antooni) [inline packages/backend/src/Application.ts:73]: I am a bit afraid that adding more and more providers will bloat the Application.ts code.
Maybe we should create one aggregate Provider which methods would take parameter {network: ethereum | optimism | etc.}

Comment (antooni) [reply packages/backend/src/Application.ts:73]: or to not break to much of current logic just add MultiNetwork provider with network parametrization

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import { TxCountRepository } from '../../peripherals/database/TxCountRepository'
+import { EthereumClient } from '../../peripherals/ethereum/EthereumClient'
+import { Clock } from '../Clock'
+
+export class RpcBlockDownloader {
+  private taskQueue = new TaskQueue<void>(() => this.update(), this.logger)
+  private blockQueue = new TaskQueue(this.getBlock.bind(this), this.logger, 100)
+
+  constructor(
+    private etheteumClient: EthereumClient,
```
Comment (antooni) [inline packages/backend/src/core/tx-count/RpcBlockDownloader.ts:13]: typo

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    private projectId: ProjectId,
+  ) {
+    this.logger = logger.for(this)
+  }
+
+  async getBlock(number: number) {
+    const block = await this.etheteumClient.getBlock(number)
+    const timestamp = new UnixTime(block.timestamp)
+
+    if (timestamp.gt(this.clock.getLastHour())) {
```
Comment (antooni) [inline packages/backend/src/core/tx-count/RpcBlockDownloader.ts:26]: I do not quite get this check, how are we able to fetch block greater than .getLastHour() ? Or is it just sanity check?

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    this.logger.info('Started')
+    return this.clock.onEveryHour(() => {
+      this.taskQueue.addIfEmpty()
+    })
+  }
+
+  async update() {
+    this.logger.info('Update started')
+
+    const missingBlocks = await this.txCountRepository.getMissingByProject(
```
Comment (antooni) [inline packages/backend/src/core/tx-count/RpcBlockDownloader.ts:48]: I assume we can have missing blocks because RPC can return an error?

File: packages/backend/src/peripherals/database/TxCountRepository.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    const knex = await this.knex()
+    const row = await knex('tx_count')
+      .where('project_id', projectId.toString())
+      .orderBy('block_number', 'desc')
+      .first()
+
+    return row ? toRecord(row) : undefined
+  }
+
+  async getMissingByProject(projectId: ProjectId) {
```
Comment (antooni) [inline packages/backend/src/peripherals/database/TxCountRepository.ts:56]: this method looks pretty complex, have you tested it with big dataset?

File: packages/backend/src/Application.ts (hunk 1 above)
Comment (michalsidzej) [reply packages/backend/src/Application.ts:73]: This is more a POC of this implementation and optimism is just an example here. So no, we will not create 10 more providers in the Application.ts code

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts (hunk 3 above)
Comment (michalsidzej) [reply packages/backend/src/core/tx-count/RpcBlockDownloader.ts:26]: This is basically why this feature is tricky. We do not hold timestamp -> blockNumber mapping in the database for all the networks. Current idea is to download all the blocks (until eth_blockNumber) and then discard all that happened after clock.getLastHour(), to avoid dealing with reorgs.

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts (hunk 4 above)
Comment (michalsidzej) [reply packages/backend/src/core/tx-count/RpcBlockDownloader.ts:48]: Yes, also the app can crash/be restarted while downloading blocks.

File: packages/backend/src/peripherals/database/TxCountRepository.ts (hunk 5 above)
Comment (michalsidzej) [reply packages/backend/src/peripherals/database/TxCountRepository.ts:56]: Onnit

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import { Logger, TaskQueue } from '@l2beat/common'
+import { ProjectId, UnixTime } from '@l2beat/types'
+
+import { TxCountRepository } from '../../peripherals/database/TxCountRepository'
+import { EthereumClient } from '../../peripherals/ethereum/EthereumClient'
+import { Clock } from '../Clock'
+
+export class RpcBlockDownloader {
```
Comment (krzkaczor) [inline packages/backend/src/core/tx-count/RpcBlockDownloader.ts:8]: This name is confusing. Primary focus of this tx is to count tx in block and insert into database

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  private taskQueue = new TaskQueue<void>(() => this.update(), this.logger)
```
Comment (krzkaczor) [inline packages/backend/src/core/tx-count/RpcBlockDownloader.ts:9]: perhpaps this could be renamed to `refreshQueue` or `updateQueue` to make it more obvious what it is

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts (hunk 3 above)
Comment (krzkaczor) [reply packages/backend/src/core/tx-count/RpcBlockDownloader.ts:26]: This deserves in code comment

File: packages/backend/src/peripherals/database/migrations/023_tx_count.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+should create a new migration file that fixes the issue.
+
+*/
+
+import { Knex } from 'knex'
+
+export async function up(knex: Knex) {
+  await knex.schema.createTable('tx_count', function (table) {
+    table.string('project_id').notNullable()
+    table.bigInteger('unix_timestamp').notNullable()
```
Comment (krzkaczor) [inline packages/backend/src/peripherals/database/migrations/023_tx_count.ts:19]: please let's switch to storing dateTime for dates. This makes working with dates in SQL easier. I did it here: https://github.com/l2beat/l2beat/pull/633/files#diff-7a11975ed53236d7e5a99aa0d987cf7f7fd2bba17c978b2a55e9cc710b0e578eR20

**Review by krzkaczor — COMMENTED** [review]
Generally speaking I like it. We ended up with a simple implementation. Tests are missing for some important things.

File: packages/backend/src/core/tx-count/RpcBlockDownloader.ts (hunk 6 above)
Comment (michalsidzej) [reply packages/backend/src/core/tx-count/RpcBlockDownloader.ts:8]: BlockTxCountUpdater?

File: packages/backend/src/core/tx-count/BlockTxCountUpdater.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+      blockNumber: block.number,
+      count: block.transactions.length,
+    })
+  }
+
+  start() {
+    this.logger.info('Started')
+    return this.clock.onEveryHour(() => {
+      this.updateQueue.addIfEmpty()
+    })
```
Comment (sz-piotr) [inline packages/backend/src/core/tx-count/BlockTxCountUpdater.ts:46]:
```suggestion
    this.updateQueue.addIfEmpty()
    return this.clock.onNewHour(() => {
      this.updateQueue.addIfEmpty()
    })
```

File: packages/backend/src/core/tx-count/BlockTxCountUpdater.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+  async update() {
+    this.logger.info('Update started')
+
+    const missingBlocks = await this.txCountRepository.getMissingByProject(
+      this.projectId,
+    )
+    for (const block of missingBlocks) {
+      this.blockQueue.addToBack(block)
+    }
```
Comment (sz-piotr) [inline packages/backend/src/core/tx-count/BlockTxCountUpdater.ts:57]: Because we call `update()` every hour thanks to the clock & queue, every hour we would add the same blocks to the queue, since they are in the queue but haven't yet been saved to the database. There should be a variable that checks that blocks up to a certain point have been queued up

File: packages/backend/src/core/tx-count/BlockTxCountUpdater.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    this.logger.info('Started')
+    return this.clock.onEveryHour(() => {
+      this.updateQueue.addIfEmpty()
+    })
+  }
+
+  async update() {
+    this.logger.info('Update started')
+
+    const missingBlocks = await this.txCountRepository.getMissingByProject(
```
Comment (sz-piotr) [inline packages/backend/src/core/tx-count/BlockTxCountUpdater.ts:52]: This approach seems to me like it would be problematic, potentially returning millions of blocks here. Maybe this could return block ranges?
Block ranges could be returned using two queries:
- find blocks for which the next block is not present
- find blocks for which the previous block is not present

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/backend/src/core/tx-count/BlockTxCountUpdater.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+import { TxCountRepository } from '../../peripherals/database/TxCountRepository'
+import { EthereumClient } from '../../peripherals/ethereum/EthereumClient'
+import { Clock } from '../Clock'
+
+export class BlockTxCountUpdater {
+  private updateQueue = new TaskQueue<void>(() => this.update(), this.logger)
+  private blockQueue = new TaskQueue(this.getBlock.bind(this), this.logger, {
+    workers: 100,
+  })
+  private queuedBlocks = new Set<number>()
```
Comment (krzkaczor) [inline packages/backend/src/core/tx-count/BlockTxCountUpdater.ts:13]: i think this implementation is acceptable for now but WDYT about adding this feature directly to TaskQueue? IMO it can be useful in other places too.

File: packages/backend/src/peripherals/database/TxCountRepository.ts [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+  async getMissingRangeByProject(projectId: ProjectId) {
+    const knex = await this.knex()
+    const maxBlock = await this.findLatestByProject(projectId)
+    if (!maxBlock) {
+      return []
+    }
+
+    const noNextBlockQuery = knex
+      .select('t1.block_number')
```
Comment (krzkaczor) [inline packages/backend/src/peripherals/database/TxCountRepository.ts:61]: I really dont think we should rely on knex query builder for such complicated queries. Plain SQL would be much more readable here IMO. @sz-piotr thoughts?

File: packages/backend/test/peripherals/database/TxCountRepository.test.ts [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        PROJECT_A_LATEST,
+      )
+      expect(await repository.findLatestByProject(PROJECT_B)).toEqual(
+        PROJECT_B_LATEST,
+      )
+    })
+  })
+
+  describe(TxCountRepository.prototype.getMissingRangeByProject.name, () => {
+    it('finds holes', async () => {
```
Comment (krzkaczor) [inline packages/backend/test/peripherals/database/TxCountRepository.test.ts:73]: this needs more tests

File: packages/backend/src/peripherals/database/TxCountRepository.ts [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    const knex = await this.knex()
+    const row = await knex('tx_count')
+      .where('project_id', projectId.toString())
+      .orderBy('block_number', 'desc')
+      .first()
+
+    return row ? toRecord(row) : undefined
+  }
+
+  async getMissingRangeByProject(projectId: ProjectId) {
```
[the remark this answers is not in the document]
Comment (krzkaczor) [reply packages/backend/src/peripherals/database/TxCountRepository.ts:53]: also, please specify what a "range" is exactly here. In this case as far as i can tell it's exclusive.

File: packages/backend/src/peripherals/database/TxCountRepository.ts (hunk 13 above)
Comment (sz-piotr) [reply packages/backend/src/peripherals/database/TxCountRepository.ts:61]: I agree, especially since we lose type safety anyway

File: packages/backend/test/peripherals/database/TxCountRepository.test.ts (hunk 14 above)
Comment (michalsidzej) [reply packages/backend/test/peripherals/database/TxCountRepository.test.ts:73]: My bad, missed the case when block 0 is missing

File: packages/backend/src/peripherals/database/migrations/023_tx_count.ts [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+If you find that something was incorrectly set up in the `up` function you
+should create a new migration file that fixes the issue.
+
+*/
+
+import { Knex } from 'knex'
+
+export async function up(knex: Knex) {
+  await knex.schema.createTable('tx_count', function (table) {
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/migrations/023_tx_count.ts:17]: This PR introduces `tx` as a shorthand for `tx` but in general in we use `transaction` everywhere. Because we dont use other shorthands too I think it's worthwile to maintain the longform. This applies to all files in this PR.

File: packages/backend/src/peripherals/database/migrations/023_tx_count.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    table.string('project_id').notNullable()
+    table.dateTime('unix_timestamp').notNullable()
+    table.integer('count').notNullable()
+    table.integer('block_number')
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/migrations/023_tx_count.ts:21]: This table has no primary key and has no indices. I think that it would be good to have a shared primary key of `project_id` and `block_number`. Also, why is `block_number` not `notNullable()`?

File: packages/backend/src/core/tx-count/BlockTxCountUpdater.ts [hunk 18]
```diff
@@ -0,0 +1,85 @@
+import { Logger, TaskQueue } from '@l2beat/common'
+import { ProjectId, UnixTime } from '@l2beat/types'
+
+import { TxCountRepository } from '../../peripherals/database/TxCountRepository'
+import { EthereumClient } from '../../peripherals/ethereum/EthereumClient'
+import { Clock } from '../Clock'
+
+export class BlockTxCountUpdater {
```
Comment (sz-piotr) [inline packages/backend/src/core/tx-count/BlockTxCountUpdater.ts:8]: I think that instead of this being called `BlockTxCountUpdater` it should be called `RpcTransactionUpdater` as most networks have blocks but not all of them use the standard ethereum rpc.

File: packages/backend/test/peripherals/database/TxCountRepository.test.ts [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+      await repository.addMany([
+        fakeTxCount({ blockNumber: 0, projectId: PROJECT_A }),
+        fakeTxCount({ blockNumber: 1, projectId: PROJECT_A }),
+        fakeTxCount({ blockNumber: 3, projectId: PROJECT_A }),
+      ])
+
+      expect(await repository.getMissingRangeByProject(PROJECT_A)).toEqual([
+        [1, 3],
```
Comment (sz-piotr) [inline packages/backend/test/peripherals/database/TxCountRepository.test.ts:81]: This is super unintuitive to me. If there are no holes we get no output. But if there is no data we also get no output. Why not have this return the following instead:
`[[-Infinity, -1], [2, 2], [4, Infinity]]`

File: packages/backend/src/core/tx-count/BlockTxCountUpdater.ts [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+      }
+    }
+
+    const lastBlock = await this.txCountRepository.findLatestByProject(
+      this.projectId,
+    )
+    const latestBlock = await this.ethereumClient.getBlockNumber()
+    let lastBlockNumber = lastBlock ? lastBlock.blockNumber : 0
+
+    while (lastBlockNumber < Number(latestBlock)) {
```
Comment (sz-piotr) [inline packages/backend/src/core/tx-count/BlockTxCountUpdater.ts:71]: If my earlier suggestion of returning ranges from -Infinity to +Infinity is to be taken into account then the logic in this file collapses to a single nested loop over ranges with Max(start, 0) and Min(end, lastBlock + 1)

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/backend/src/peripherals/database/migrations/023_tx_count.ts (hunk 17 above)
Comment (michalsidzej) [reply packages/backend/src/peripherals/database/migrations/023_tx_count.ts:21]: So the idea was primarily that some of the projects don't give us block_number (ex. all the StarkEx implementations)

File: packages/backend/src/core/tx-count/BlockTxCountUpdater.ts [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+      for (let i = range[0] + 1; i < range[1]; i++) {
+        this.queueBlock(i)
+      }
+    }
+
+    const lastBlock = await this.txCountRepository.findLatestByProject(
+      this.projectId,
+    )
+    const latestBlock = await this.ethereumClient.getBlockNumber()
+    let lastBlockNumber = lastBlock ? lastBlock.blockNumber : 0
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/core/tx-count/BlockTxCountUpdater.ts]: That's a good question, WYT @krzkaczor? Block 0 will be full of artificial transactions that bring no useful info for us.

[the remark this answers is not in the document]
Comment (krzkaczor) [reply packages/backend/src/core/tx-count/BlockTxCountUpdater.ts]: I'd say, gather everything in the database. If it's gonna be really odd data, we will filter it out with a simple SQL query.

File: packages/backend/src/peripherals/database/TxCountRepository.ts [hunk 22]
```diff
@@ -50,59 +55,51 @@ export class TxCountRepository extends BaseRepository {
...
+        WHERE b2.block_number IS NULL) AS no_next
+      ORDER BY block_number ASC
+    `)) as unknown as RawBlockNumberQueryResult
 
     // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-    const ranges: [TxCountRow, TxCountRow][] = _.zip(noNext, noPrev)
+    const noPrev = (await knex.raw(`
+      WITH 
+          project_blocks AS (
+            SELECT * FROM tx_count WHERE project_id='${projectId.toString()}'
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/TxCountRepository.ts]: Potential sql injection vulnerability. Use bindings instead of string manipulation.
https://knexjs.org/guide/raw.html#raw-parameter-binding

File: packages/backend/src/peripherals/database/TxCountRepository.ts [hunk 23]
```diff
@@ -50,59 +55,51 @@ export class TxCountRepository extends BaseRepository {
...
-      .with(
-        'no_prev',
-        knex.raw('? EXCEPT ?', [noPrevBlockQuery, noNextBlockQuery]),
-      )
-      .from('no_prev')
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const noNext = (await knex.raw(`
+      WITH 
+        project_blocks AS (
+          SELECT * FROM tx_count WHERE project_id='${projectId.toString()}'
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/TxCountRepository.ts]: bindings

File: packages/backend/src/peripherals/database/TxCountRepository.ts [hunk 24]
```diff
@@ -50,59 +55,51 @@ export class TxCountRepository extends BaseRepository {
...
-      .from('no_next')
-
-    const noPrev = await knex
-      .select()
-      .with(
-        'no_prev',
-        knex.raw('? EXCEPT ?', [noPrevBlockQuery, noNextBlockQuery]),
-      )
-      .from('no_prev')
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/TxCountRepository.ts]: Is this still needed if you cast the expression?

File: packages/backend/test/core/tx-count/BlockTxCountUpdater.test.ts [hunk 25]
```diff
@@ -20,7 +20,7 @@ describe(BlockTxCountUpdater.name, () => {
       })
       const txCountRepository = mock<TxCountRepository>({
         findLatestByProject: async () => fakeTxCount({ blockNumber: 4 }),
-        getMissingRangeByProject: async () => [[1, 3]],
+        getMissingRangesByProject: async () => [[1, 3]],
```
Comment (sz-piotr) [inline packages/backend/test/core/tx-count/BlockTxCountUpdater.test.ts]: This is a value that this function will never return

File: packages/backend/test/core/tx-count/BlockTxCountUpdater.test.ts [hunk 26]
```diff
@@ -53,7 +53,7 @@ describe(BlockTxCountUpdater.name, () => {
...
```
Comment (sz-piotr) [inline packages/backend/test/core/tx-count/BlockTxCountUpdater.test.ts]: This is a value that this function will never return

File: packages/backend/test/core/tx-count/BlockTxCountUpdater.test.ts (hunk 25 above)
Comment (michalsidzej) [reply packages/backend/test/core/tx-count/BlockTxCountUpdater.test.ts]: This is outdated, fixed in next commit

File: packages/backend/test/core/tx-count/BlockTxCountUpdater.test.ts (hunk 26 above)
Comment (michalsidzej) [reply packages/backend/test/core/tx-count/BlockTxCountUpdater.test.ts]: Outdated

File: packages/backend/src/peripherals/database/migrations/023_tx_count.ts (hunk 17 above)
Comment (michalsidzej) [reply packages/backend/src/peripherals/database/migrations/023_tx_count.ts]: Thinking of this rn, we should probably have 2 tables, where 1 is only for raw block data and the second would hold aggregated data, ie for StarkEx

File: packages/backend/test/peripherals/database/TransactionCountRepository.test.ts [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+  const PROJECT_B = ProjectId('project-b')
+
+  beforeEach(async () => {
+    await repository.deleteAll()
+  })
+
+  describe(
+    TransactionCountRepository.prototype.getMissingRangesByProject.name,
+    () => {
+      it('finds holes', async () => {
```
Comment (sz-piotr) [inline packages/backend/test/peripherals/database/TransactionCountRepository.test.ts:25]: There is still an edge case test missing for a completely empty repository. We want it to return [-Infinity, Infinity]

**Review by sz-piotr — APPROVED** [review]

File: packages/backend/src/core/tx-count/BlockTxCountUpdater.ts (hunk 12 above)
Comment (michalsidzej) [reply packages/backend/src/core/tx-count/BlockTxCountUpdater.ts]: @krzkaczor  Should I add this as a task to backlog?

**Review by antooni — APPROVED** [review]

**Review by krzkaczor — APPROVED** [review]
